### PR TITLE
Use TemporaryDirectory vs string for temp directories

### PR DESCRIPTION
**Pull request by abaptiste**
This change addresses https://github.com/envoyproxy/envoy-perf/issues/90.  When creating temporary directories we use a `TemporaryDirectory` object instead of a string.

Signed-off-by: abaptiste <abaptiste@users.noreply.github.com>

cc: @mum4k , @landesherr

### CODE REVIEW

File: salvo/src/lib/cmd_exec.py [hunk 1]
```diff
@@ -35,18 +36,32 @@ def run_command(cmd: str, parameters: CommandParameters) -> str:
...
   try:
     log.debug(f"Executing command: [{cmd}] with args [{parameters._asdict()}]")
     cmd_array = shlex.split(cmd)
-    output = subprocess.check_output(
-        cmd_array, stderr=subprocess.STDOUT, **parameters._asdict())
 
-    if isinstance(output, bytes):
-      output = output.decode('utf-8').strip()
+    subprocess.check_call(cmd_array, stdout=tmpfile, stderr=tmpfile,
+                          **parameters._asdict())
```
Comment (mum4k) [inline salvo/src/lib/cmd_exec.py:49]: (nit) Can we use `params` here?

**params that is.

File: salvo/src/lib/source_tree.py [hunk 2]
```diff
@@ -164,8 +166,31 @@ def get_source_directory(self) -> str:
     """
     self._validate()
 
-    return  self._source_repo.source_path \
-                 if self._source_repo.source_path else self._tempdir
+    return self._build_dir.name
+
+  def copy_source_directory(self) -> bool:
```
Comment (mum4k) [inline salvo/src/lib/source_tree.py:171]: Are these changes also relevant to #90? If not, can we move them to a separate PR to keep this one focused?

File: salvo/src/lib/source_tree.py [hunk 3]
```diff
@@ -54,7 +56,11 @@ def __init__(self, source_repo: proto_source.SourceRepository) -> None:
     if not home_dir.startswith(constants.SALVO_TMP):
       home_dir = constants.SALVO_TMP
 
-    self._tempdir = file_ops.get_random_dir(home_dir)
+    # This is the destination directory where the source is copied or cloned
+    # If the source_repo has a path, we copy from that path into the _build_dir.
+    # If the source_repo has a url, we clone from that url into the _build_dir.
+    self._build_dir = file_ops.get_random_dir(home_dir)
```
Comment (mum4k) [inline salvo/src/lib/source_tree.py:62]: Since we are storing this in an attribute, we aren't really addressing #90, since nothing will destroy the temporary directory after we use it like this. However since this is the build directory we are producing, do we actually want to destroy it after the execution? Or do we want to keep it around for debugging?

If we do want to destroy it, we should ensure that the `__exit__()` method of the created temp directory gets called at some point.

File: salvo/src/lib/cmd_exec.py [hunk 4]
```diff
@@ -35,18 +36,32 @@ def run_command(cmd: str, parameters: CommandParameters) -> str:
...
 
-    log.debug(f"Returning output: [{output}]")
   except subprocess.CalledProcessError as process_error:
     log.error(f"Unable to execute [{cmd}]: {process_error}")
     raise
 
+  finally:
+    tmpfile.flush()
```
Comment (mum4k) [inline salvo/src/lib/cmd_exec.py:56]: Is using the temp instead of a string here an improvement? Looks like we have to do some number of operations to get this output now (flush, seek, read). What do we gain by switching this into a file instead of a string. Is this relevant to #90 which was talking about automatic cleanup? Looks like the original solution was already performing cleanup, since it wasn't writing to any files, while the current version seems to be leaving the temp file behind.

Or are we trying to achieve something other than what is listed in #90?

**Review by mum4k — CHANGES_REQUESTED** [review]

File: salvo/src/lib/cmd_exec.py (hunk 1 above)
Comment (abaptiste) [reply salvo/src/lib/cmd_exec.py:49]: With this change I get:
```
TypeError: check_call() argument after ** must be a mapping, not CommandParameters
```

File: salvo/src/lib/source_tree.py (hunk 2 above)
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:171]: I can pull this out of this PR

File: salvo/src/lib/source_tree.py (hunk 3 above)
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:62]: I know that when the builder objects go out of scope the temporary directories get cleaned up.   I had to chase a few "File Not Found" errors due to the TemporaryDirectory getting removed.

I know I have to do some work ensuring that various artifacts are cleaned up after benchmarks are executed. We could define `__del__` in the source_tree to ensure that these objects are closed and purged, but that's a separate effort I haven't started digging into yet.

File: salvo/src/lib/cmd_exec.py (hunk 4 above)
Comment (abaptiste) [reply salvo/src/lib/cmd_exec.py:56]: The output from some benchmarks is rather large.  This seems to work more consistently in capturing the output.

I kept this in since it did alert me to a few TemporaryDirectory objects getting prematurely reaped while addressing https://github.com/envoyproxy/envoy-perf/issues/90.

I'd prefer to keep it.

File: salvo/src/lib/cmd_exec.py (hunk 1 above)
Comment (mum4k) [reply salvo/src/lib/cmd_exec.py:49]: Thank you trying it out.

File: salvo/src/lib/cmd_exec.py (hunk 4 above)
Comment (mum4k) [reply salvo/src/lib/cmd_exec.py:56]: Thank you for helping me understand this. Since this qualifies as a tricky (or non-obvious) code, can we add a inline comment explaining the need for this construct? This will help future maintainers.

Comment (abaptiste) [reply salvo/src/lib/cmd_exec.py:56]: Sure thing. Let me update the PR.

File: salvo/src/lib/source_tree.py (hunk 3 above)
Comment (mum4k) [reply salvo/src/lib/source_tree.py:62]: Thank you, tracking it as a separate effort is a good idea.

**Review by mum4k — APPROVED** [review]
